Automattic/vip-go-mu-plugins
# Add cache to the months dropdown filter for posts

## Description

The months dropdown filter on the listing pages can result in a slow request when the database is large.

I have monitored around 500ms for 500k+ posts. The request to generate the dropdown is here: https://github.com/WordPress/WordPress/blob/e89f6692745937af4bfe5acbf8c5918f393bbc77/wp-admin/includes/class-wp-list-table.php#L609-L620

As backend requests are not cached this is happening on each page load.

The filter required for this PR to work is not yet implemented in WordPress core. [An issue has been opened](https://core.trac.wordpress.org/ticket/51660) and **is mandatory for this PR**.

This PR is hooking on this potential new filter and save the months dropdown in a transient for each post type. I have also added a hook on `save_post` to bust the cache when necessary.

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

1. Go to a listing page e.g.: `/wp-admin/edit.php` or a custom post type listing page (`/wp-admin/edit.php?post_type=myCustomPostType`)
2. Once the page has been loaded, a transient should have been added to the `wp_options` table named `_transient_wpcom_vip_edit_months_array_post` or `_transient_wpcom_vip_edit_months_array_myCustomPostType` saving the months dropdown.
3. Reload the listing page
4. Check with the Query monitor plugin that the request is no more targeting the database but the transient.
5. Extra: try to bust the cache by removing the only post for a specific month
